tools/keyvaluestore.go

```diff
@@ -0,0 +1,201 @@
+package tools
+
+import (
+	"encoding/gob"
+	"fmt"
+	"io"
+	"os"
+	"sync"
+)
+
+// KeyValueStore provides an in-memory key/value store which is persisted to
+// a file. The file handle itself is not kept locked for the duration; it is
+// only locked during load and save, to make it concurrency friendly. When
+// saving, the store uses optimistic locking to determine whether the db on disk
+// has been modified by another process; in which case it loads the latest
+// version and re-applies modifications made during this session. This means
+// the Lost Update db concurrency issue is possible; so don't use this if you
+// need more DB integrity than Read Committed isolation levels.
+type KeyValueStore struct {
+	mu       sync.RWMutex
+	filename string
+	log      []keyValueChange
+
+	// This is the persistent data
+	// version for optimistic locking, this field is incremented with every Save()
+	version int64
+	db      map[string]interface{}
+}
+
+type keyValueOperation int
```

Comment:
Yeah, although because of data alignment, in practice you don't actually save anything unless you have an array of them :)